arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2890,6 +2915,297 @@ void RestReplicationHandler::handleCommandLoggerTickRanges() {
+  std::size_t const rangeMin = static_cast<std::size_t>(ctx.collection->minRevision());
+  containers::RevisionTree tree(maxDepth, rangeMin);
+
+  while (it.hasMore()) {
+    tree.insert(it.revision(), TRI_FnvHashPod(it.revision())/*it.document().hashString()*/);
+    it.next();
+  }
+
+  std::string result = tree.serialize();
+
+  resetResponse(rest::ResponseCode::OK);
+  _response->setContentType("application/octet-stream");
+  _response->addRawPayload(VPackStringRef(result.data(), result.size()));
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief return the requested revision ranges for a given collection, if
+///        available
+/// @response VPackObject, containing
+///           * ranges, VPackArray of VPackArray of revisions
+///           * resume, optional, if response is chunked; revision resume
+///                     point to specify on subsequent requests
+//////////////////////////////////////////////////////////////////////////////
+
+void RestReplicationHandler::handleCommandRevisionRanges() {
+  RevisionOperationContext ctx;
+  if (!prepareRevisionOperation(ctx)) {
+    return;
+  }
+
+  bool success = false;
+  std::size_t current = 0;
+  VPackSlice const body = this->parseVPackBody(success);
+  if (!success) {
+    // error already created
+    return;
+  }
+  bool badFormat = !body.isArray();
+  if (!badFormat) {
+    for (std::size_t i = 0; i < body.length(); ++i) {
+      if (!body.at(i).isArray() || body.at(i).length() != 2 ||
+          !body.at(i).at(0).isNumber() || !body.at(i).at(1).isNumber()) {
+        badFormat = true;
+        break;
+      }
+      if (body.at(i).at(0).getNumber<std::size_t>() <= ctx.resume ||
+          (i > 0 && body.at(i - 1).at(1).getNumber<std::size_t>() < ctx.resume)) {
+        current = std::max(current, i);
+      }
+    }
+  }
+  if (badFormat) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                  "body needs to be an array of pairs of revisions");
+    return;
+  }
+
+  RevisionReplicationIterator& it =
+      *static_cast<RevisionReplicationIterator*>(ctx.iter.get());
+  it.seek(ctx.resume);
+
+  std::size_t constexpr limit = 64 * 1024;  // ~2MB data
+  std::size_t total = 0;
+  VPackBuilder response;
+  {
+    VPackObjectBuilder obj(&response);
+    TRI_voc_rid_t resumeNext = ctx.resume;
+    VPackSlice range = body.at(current);
+
+    TRI_ASSERT(response.isOpenObject());
+    response.add(VPackValue("ranges"));
+    response.openArray();
+    TRI_ASSERT(response.isOpenArray());
+    bool subOpen = false;
+    while (total < limit && current < body.length()) {
+      if (!it.hasMore() || it.revision() <= range.at(0).getNumber<std::size_t>() ||
+          it.revision() > range.at(1).getNumber<std::size_t>() ||
+          (!subOpen && it.revision() <= range.at(1).getNumber<std::size_t>())) {
+        it.seek(std::max(ctx.resume, range.at(0).getNumber<std::size_t>()));
```

Comment:
seeking is probably expensive in rocksdb, why don't we compare the current key before we seek?